spherical_splines/SHTOOLS/man/man1/makegridglq.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAKEGRIDGLQ 1"
.TH MAKEGRIDGLQ 1 "2009-08-21" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "MakeGridGLQ"
.IX Header "MakeGridGLQ"
.IP "MakeGridGLQ \-" 15
.IX Item "MakeGridGLQ -"
Create a 2D map from a set of spherical harmonic coefficients sampled on the Gauss-Legendre quadrature nodes.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 MakeGridGLQ (" 25
.IX Item "SUBROUTINE MakeGridGLQ ("
\&\s-1GRIDGLQ\s0, \s-1CILM\s0, \s-1LMAX\s0, \s-1PLX\s0, \s-1ZERO\s0, \s-1NORM\s0, \s-1CSPHASE\s0, \s-1LMAX_CALC\s0 )
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1GRIDGLQ\s0(\s-1LMAX+1\s0, 2*LMAX+1), \s-1CILM\s0(2,\ *,\ *)
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
\&\s-1LMAX\s0
.IP "REAL*8, \s-1OPTIONAL\s0" 19
.IX Item "REAL*8, OPTIONAL"
\&\s-1PLX\s0(\s-1LMAX+1\s0, (\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2), \s-1ZERO\s0(\s-1LMAX+1\s0)
.IP "\s-1INTEGER\s0, \s-1OPTIONAL\s0" 19
.IX Item "INTEGER, OPTIONAL"
\&\s-1NORM\s0, \s-1CSPHASE\s0, \s-1LMAX_CALC\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIMakeGridGLQ\fR will create a 2\-dimensional map from a set of input spherical harmonic coefficients sampled on the Gauss-Legendre quadrature nodes. This is the inverse of the routine \fISHExpandGLQ\fR. The latitudinal nodes correspond to the zeros of the Legendre polynomial of degree \s-1LMAX+1\s0, and the longitudinal nodes are equally spaced with an interval of 360/(2*LMAX+1) degrees. When evaluating the function, the maximum spherical harmonic degree that is considered is the minimum of \s-1LMAX\s0, the size of \s-1CILM\-1\s0, or \s-1LMAX_CALC\s0 (if specified).
.PP
The employed spherical harmonic normalization and Condon-Shortley phase convention can be set by the optional arguments \s-1NORM\s0 and \s-1CSPHASE\s0; if not set, the default is to use geodesy 4\-pi normalized harmonics that exclude the Condon-Shortley phase of (\-1)^m.
.PP
The reconstruction of the spherical harmonic function can be speeded up by precomputing the Legendre functions on the Gauss-Legendre quadrature nodes in the routine \fIPreCompute\fR. However, given that this array contains on the order of LMAX**3 entries, this is only feasible for moderate values of \s-1LMAX\s0.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1GRIDGLQ\s0" 11
.IX Item "GRIDGLQ"
(output) REAL*8, \s-1DIMENSION\s0(\s-1LMAX+1\s0, 2*LMAX+1)
.Sp
A 2D map of the function sampled on the Gauss-Legendre quadrature nodes.
.IP "\s-1CILM\s0" 11
.IX Item "CILM"
(input) REAL*8, \s-1DIMENSION\s0 (2, *, *)
.Sp
The real spherical harmonic coefficients of the function. When evaluating the function, the maximum spherical harmonic degree considered is the minimum of \s-1LMAX\s0, the size of \s-1CILM\s0(:,*,*) \- 1, or \s-1LMAX_CALC\s0 (if specified). The coefficients C1lm and C2lm refer to the \*(L"cosine\*(R" (Clm) and \*(L"sine\*(R" (Slm) coefficients, respectively, with Clm = \s-1CILM\s0(1, l+1, m+1) and Slm = \s-1CILM\s0(2, l+1, m+1). Alternatively, C1lm and C2lm correspond to the positive and negative order coefficients, respectively.
.IP "\s-1LMAX\s0" 11
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum spherical harmonic bandwidth of the function. This determines the sampling nodes of the output grid.
.IP "\s-1PLX\s0" 11
.IX Item "PLX"
(input) REAL*8, \s-1OPTIONAL\s0, \s-1DIMENSION\s0 (\s-1LMAX+1\s0, (\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2)
.Sp
An array of the associated Legendre functions calculated at the Gauss-Legendre quadrature nodes. These are determined from a call to \fIPreCompute\fR. Either \s-1PLX\s0 or \s-1ZERO\s0 must be present, but not both.
.IP "\s-1ZERO\s0" 11
.IX Item "ZERO"
(input) REAL*8, \s-1OPTIONAL\s0, \s-1DIMENSION\s0 (\s-1LMAX+1\s0)
.Sp
The nodes used in the Gauss-Legendre quadrature over latitude, calculated by a call to \fIPreCompute\fR.  Either \s-1PLX\s0 or \s-1ZERO\s0 must be present, but not both.
.IP "\s-1NORM\s0" 11
.IX Item "NORM"

spherical_splines/SHTOOLS/src/MakeGridGLQ.f95
subroutine MakeGridGLQ(gridglq, cilm, lmax, plx, zero, norm, csphase, lmax_calc)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	Given the Spherical Harmonic coefficients CILM, this subroutine
!	will evalate these coefficients on a grid with equal spacing in
!	longitude and with latitude points appropriate for Gauss-Legendre
!	quadrature integrations. Note that this is done using FFTs for each degree of
!	each latitude band. The grid spacing is determined by the spherical harmonic
!	bandwidth LMAX. Nevertheless, the coefficients can be evaluated up to
!	smaller spherical harmonic degree by specifying the optional parameter
!	LMAX_CALC.
!
!	The optional array PLX contains precomputed associated legendre functions
!	evaluated on the Gauss-Legendre quadrature nodes (obtained from PreCompute)
!	and should not be precomputed when memory is an issue (i.e., lmax>360).
!	If PLX is not present, the Legendre functions are computed on the fly
!	using the scaling methodolgy presented in Holmes and Featherston (2002). 
!	When NORM=1,2 or 4, these are accurate to degree 2800. When NORM=3, the
!	routine is only stable to about degree 15!
!
!	Calling Parameters
!		IN
!			cilm		Input spherical harmonic coefficients with 
!					dimensions (2, LMAX+1, LMAX+1).
!			lmax		Maximum spherical harmonic degree used in the expansion.
!					This value determines the grid spacing of the output function.
!		OUT
!			gridglq		Gridded data of the spherical harmonic
!					coefficients CILM with dimensions (LMAX+1 , 2*LMAX+1). 
!					The first index (latitude) corresponds to the 
!					Gauss points, and the second index corresponds to 
!					360*(k-1)/nlong = 360*(k-1)/(2*LMAX +1). 
!		OPTIONAL (IN)
!			plx:		Input array of Associated Legendre Polnomials computed
!					at the Gauss points (determined from a call to
!					PreCompute). If this is not included, then the optional
!					array ZERO MUST be inlcuded.
!			zero		Array of dimension lmax+1 that contains the latitudinal
!					gridpoints used in the Gauss-Legendre quadrature integration
!					scheme. Only needed if PLX is not included.
!			norm		Normalization to be used when calculating Legendre functions
!						(1) "geodesy" (default)
!						(2) Schmidt
!						(3) unnormalized
!						(4) orthonormalized
!			csphase	1: Do not include the phase factor of (-1)^m
!					-1: Apply the phase factor of (-1)^m.
!			lmax_calx	The maximum spherical harmonic degree to evaluate the coefficients
!					up to.
!
!	Notes:
!		1.	If lmax is greater than the the maximum spherical harmonic
!			degree of the input file, then this file will be ZERO PADDED!
!			(i.e., those degrees after lmax are assumed to be zero).
!		2.	Latitudes are geocentric latitude.
!
!	Dependencies:	FFTW3, NGLQSH, CSPHASE_DEFAULT
!
!	Written by Mark Wieczorek 2004.
!
!	September 3, 2005. 	Modifed so that the array plx is now optional.
!	September 26, 2005: 	Added optional argument NORM.
!	October 16, 2006. 	The Legendre functions are now computed within this program
!				during the summations over l and m. This leads to an increase 
!				in speed of about a factor of 2.
!	July 22, 2007. 		Added optional parameter LMAX_CALC for evaluating coefficients up
!				to a maximum degree smaller than LMAX.
!	November 21, 2011	Fixed problem where saved variables used in Plm recursion were not recalculated
!				if NORM changed from one call to the next (with the same value of N).
!
!	Copyright (c) 2005-2011, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use FFTW3
	use SHTOOLS, only: NGLQSH, CSPHASE_DEFAULT
	
	implicit none
	
	real*8, intent(in) :: 	cilm(:,:,:)
	real*8, intent(in), optional :: plx(:,:), zero(:)
	real*8, intent(out) ::	gridglq(:,:)
	integer, intent(in) :: 	lmax
	integer, intent(in), optional :: norm, csphase, lmax_calc
	integer :: 		l, m, i, nlat, nlong, l1, m1, lmax_comp, phase, i_s, astat(4), lnorm, &
				k, kstart
	real*8 :: 		grid(2*lmax+1), pi, coef0, coef0s, scalef, rescalem, u, p, pmm, pm1, pm2, z
	complex*16 ::		coef(lmax+1), coefs(lmax+1)
	integer*8 ::		plan
	real*8, save, allocatable ::	f1(:), f2(:), sqr(:), symsign(:)
	integer, save ::	lmax_old=0, norm_old = 0


	if (size(cilm(:,1,1)) < 2) then
		print*, "Error --- MakeGridGLQ"
		print*, "CILM must be dimensioned as (2, *, *)."
		print*, "Input dimension is ", size(cilm(:,1,1)), size(cilm(1,:,1)), size(cilm(1,1,:))
		stop
	elseif (size(gridglq(1,:)) < 2*lmax+1 .or. size(gridglq(:,1)) < lmax+1 ) then
		print*, "Error --- MakeGridGLQ"
		print*, "GRIDGLQ must be dimensioned as (LMAX+1, 2*LMAX+1) where LMAX is ", lmax
		print*, "Input array is dimensioned ", size(gridglq(:,1)), size(gridglq(1,:))
		stop
	endif

	if (present(plx)) then
		if (size(plx(:,1)) < lmax+1 .or. size(plx(1,:)) < (lmax+1)*(lmax+2)/2) then
			print*, "Error --- MakeGridGLQ"
			print*, "PLX must be dimensioned as (LMAX+1, (LMAX+1)*(LMAX+2)/2) where LMAX is ", lmax
			print*, "Input array is dimensioned as ", size(plx(:,1)), size(plx(1,:))
			stop
		endif
	elseif (present(zero)) then
		if (size(zero) < lmax+1) then
			print*, "Error --- MakeGridGLQ"
			print*, "ZERO must be dimensioned as (LMAX+1) where LMAX is ", lmax
			print*, "Input array is dimensioned ", size(zero)
			stop
		endif
	else
		print*, "Error --- MakeGridGLQ"
		print*, "Either PLX or ZERO must be specified."
		stop
	endif
	
	if (present(norm)) then
		if (norm > 4 .or. norm < 1) then
			print*, "Error - MakeGridGLQ"
			print*, "Parameter NORM must be 1 (geodesy), 2 (Schmidt), 3 (unnormalized), or 4 (orthonormalized)."
			print*, "Input value is ", norm
			stop
		endif
		
		lnorm = norm
		
	else
		lnorm = 1
	endif
	
	if (present(csphase)) then
     		if (csphase /= -1 .and. csphase /= 1) then
     			print*, "Error --- MakeGridGLQ"
     			print*, "CSPHASE must be 1 (exclude) or -1 (include)."
     			stop
     		else
     			phase = csphase
     		endif
     	else
     		phase = CSPHASE_DEFAULT
     	endif
     	
	
	pi = acos(-1.0d0)
	
	nlong = 2*lmax + 1
	
	nlat = NGLQSH(lmax)	! nlat = lmax+1
	
	scalef = 1.0d-280
	
	if (present(lmax_calc)) then
		lmax_comp = min(lmax, size(cilm(1,1,:))-1, size(cilm(1,:,1))-1, lmax_calc)
	else
		lmax_comp = min(lmax, size(cilm(1,1,:))-1, size(cilm(1,:,1))-1)
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate recursion constants used in computing the Legendre polynomials
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	if ( (lmax_comp /= lmax_old .or. lnorm /= norm_old) .and. .not. present(plx)) then
		
		if (allocated(sqr)) deallocate(sqr)
		if (allocated(f1)) deallocate(f1)
		if (allocated(f2)) deallocate(f2)
		if (allocated(symsign)) deallocate(symsign)
		
		allocate(sqr(2*lmax_comp+1), stat=astat(1))
		allocate(f1((lmax_comp+1)*(lmax_comp+2)/2), stat=astat(2))
		allocate(f2((lmax_comp+1)*(lmax_comp+2)/2), stat=astat(3))
		allocate(symsign((lmax_comp+1)*(lmax_comp+2)/2), stat=astat(4))
		
		if (astat(1) /= 0 .or. astat(2) /= 0 .or. astat(3) /= 0 .or. astat(4) /= 0) then
			print*, "MakeGridGLQ --- Error"
			print*, "Problem allocating arrays SQR, F1, F2, or SYMSIGN", astat(1), astat(2), astat(3), astat(4)
			stop
		endif
		
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		! 	Calculate signs used for symmetry of Legendre functions about equator
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		
		k = 0
		do l = 0, lmax_comp, 1
			do m = 0, l, 1
				k = k + 1
				if (mod(l-m,2) == 0) then
					symsign(k) = 1.0d0
				else
					symsign(k) = -1.0d0
				endif
			enddo
		enddo
			
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		!	Precompute square roots of integers that are used several times.
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
		do l=1, 2*lmax_comp+1
			sqr(l) = sqrt(dble(l))
		enddo

		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		! 	Precompute multiplicative factors used in recursion relationships
		! 		P(l,m) = x*f1(l,m)*P(l-1,m) - P(l-2,m)*f2(l,m)
		!		k = l*(l+1)/2 + m + 1
		!	Note that prefactors are not used for the case when m=l as a different 
		!	recursion is used. Furthermore, for m=l-1, Plmbar(l-2,m) is assumed to be zero.
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
		k = 1
		
		select case(lnorm)
		
			case(1,4)
				
				if (lmax_comp /= 0) then
					k = k + 1
					f1(k) = sqr(3)
					f2(k) = 0.0d0
					k = k + 1
				endif
				
				do l=2, lmax_comp, 1
					k = k + 1
					f1(k) = sqr(2*l-1) * sqr(2*l+1) / dble(l)
					f2(k) = dble(l-1) * sqr(2*l+1) / sqr(2*l-3) / dble(l)
					do m=1, l-2, 1
						k = k+1
						f1(k) = sqr(2*l+1) * sqr(2*l-1) / sqr(l+m) / sqr(l-m)
                				f2(k) = sqr(2*l+1) * sqr(l-m-1) * sqr(l+m-1) &
                  				 	/ sqr(2*l-3) / sqr(l+m) / sqr(l-m) 
					enddo
					k = k+1
					f1(k) = sqr(2*l+1) * sqr(2*l-1) / sqr(l+m) / sqr(l-m)
                			f2(k) = 0.0d0
					k = k + 1
				enddo
			
			case(2)
				
				if (lmax_comp /= 0) then
					k = k + 1
					f1(k) = 1.0d0
					f2(k) = 0.0d0
					k = k + 1
				endif
				
				do l=2, lmax_comp, 1
					k = k + 1
					f1(k) = dble(2*l-1) /dble(l)
					f2(k) = dble(l-1) /dble(l)
					do m=1, l-2, 1
						k = k+1
						f1(k) = dble(2*l-1) / sqr(l+m) / sqr(l-m)
                  				f2(k) = sqr(l-m-1) * sqr(l+m-1) / sqr(l+m) / sqr(l-m)
					enddo
					k = k+1
					f1(k) = dble(2*l-1) / sqr(l+m) / sqr(l-m)
                  			f2(k) = 0.0d0
					k = k + 1
				enddo
			
			case(3)		
			
				do l=1, lmax_comp, 1
					k = k + 1
					f1(k) = dble(2*l-1) /dble(l)
					f2(k) = dble(l-1) /dble(l)
					do m=1, l-1, 1
						k = k+1
						f1(k) = dble(2*l-1) / dble(l-m)
                  				f2(k) = dble(l+m-1) / dble(l-m)
					enddo
					k = k + 1
				enddo

		end select
	
		lmax_old = lmax_comp
		norm_old = lnorm
	
	endif	
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Do special case of lmax_comp = 0
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	if (lmax_comp == 0) then
	
		select case(lnorm)
			case(1,2,3);	pm2 = 1.0d0
			case(4);	pm2 = 1.0d0 / sqrt(4.0d0*pi)
		end select
		
		gridglq(1:nlat, 1:nlong) = cilm(1,1,1) * pm2
	
		return
	
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Determine Cilms, one l at a time I by integrating over all 
	!	latitudes using Gauss-Legendre Quadrature. When PLX is not 
	!	present, the Legendre functions are computed on the fly 
	!	during the summations over l and m. These are scaled using
	!	the methodology of Holmesand Featherstone (2002), with the
	!	exception of the m=0 terms that do not need to be scaled 
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	call dfftw_plan_dft_c2r_1d(plan, nlong, coef, grid, FFTW_MEASURE)	! create generic plan
	
	if (present(plx)) then
	
		do i = 1, nlat
			coef0 = 0.0d0
			coef = dcmplx(0.0d0,0.0d0)
			! This summation order is intended to add the smallest terms first
			do l = lmax_comp, 0, -1
				l1 = l+1
				k =  l1*l/2 + 1	! m=0
				coef0 = coef0 + cilm(1,l1,1) * plx(i,k)
				do m = 1, l, 1
					m1 = m+1
					k = l1*l/2 + m1
					coef(m1) = coef(m1) + dcmplx(cilm(1,l1,m1), - cilm(2,l1,m1)) &
						* plx(i,k)/2.0d0
				enddo
			enddo
		
			coef(1) = dcmplx(coef0, 0.0d0)
			
               		call dfftw_execute(plan)	! take fourier transform
                
			gridglq(i,1:nlong) = grid(1:nlong)
		enddo
		
	else

		do i=1, (nlat+1)/2
			
			coef = dcmplx(0.0d0,0.0d0)
			coef0 = 0.0d0
		
			if ( i==(nlat+1)/2 .and. mod(nlat,2) /=0) then		! This latitude is the equator; z=0, u=1
				
				u = 1.0d0
				
				select case(lnorm)
					case(1,2,3);	pm2 = 1.0d0
					case(4);	pm2 = 1.0d0 / sqrt(4.0d0*pi)
				end select

				coef0 = coef0 + cilm(1,1,1) * pm2
				
				k = 2
				
				do l=2, lmax_comp, 2
					l1 = l+1
					k = k + l
					p = - f2(k) * pm2
					pm2 = p
					coef0 = coef0 + cilm(1,l1,1) * p
					k = k + l + 1
				enddo
				
				select case(lnorm)
					case(1,2);	pmm = sqr(2) * scalef
					case(3);	pmm = scalef
					case(4);	pmm = sqr(2) * scalef / sqrt(4.0d0*pi)
				end select
								
				rescalem = 1.0d0/scalef
				kstart = 1
			
				do m = 1, lmax_comp-1, 1
				
					m1 = m + 1
					kstart = kstart + m + 1
					
					select case(lnorm)
						case(1,4)
							pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
							pm2 = pmm
						case(2)
							pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
							pm2 = pmm / sqr(2*m+1)
						case(3)
							pmm = phase * pmm * dble(2*m-1)
							pm2 = pmm
					end select
					
					coef(m1) = coef(m1) + dcmplx(cilm(1,m1,m1), &
							- cilm(2,m1,m1)) * pm2
					
					k = kstart+m+1
	   				
					do l = m+2, lmax_comp, 2
						l1 = l+1
						k = k + l
						p = - f2(k) * pm2
						coef(m1) = coef(m1) + dcmplx(cilm(1,l1,m1), &
							- cilm(2,l1,m1)) * p
						pm2 = p
						k = k + l + 1
					enddo
										
				enddo			
				
           			select case(lnorm)
            				case(1,4);	pmm = phase * pmm * sqr(2*lmax_comp+1) / sqr(2*lmax_comp)
            				case(2);	pmm = phase * pmm / sqr(2*lmax_comp)
            				case(3);	pmm = phase * pmm * dble(2*lmax_comp-1)
        			end select
          			
        			coef(lmax_comp+1) = coef(lmax_comp+1) + dcmplx(cilm(1,lmax_comp+1,lmax_comp+1), &
							- cilm(2,lmax_comp+1,lmax_comp+1)) * pmm
		
				coef(1) = dcmplx(coef0,0.0d0)
				coef(2:lmax_comp+1) = coef(2:lmax_comp+1) * rescalem /2.0d0
               			call dfftw_execute(plan)	! take fourier transform
                
				gridglq(i,1:nlong) = grid(1:nlong)
			
			else
			
				z = zero(i)
				u = sqrt( (1.0d0-z) * (1.0d0+z) )
				
				i_s = nlat+1-i

				coefs = dcmplx(0.0d0,0.0d0)
				coef0s = 0.0d0
				
				select case(lnorm)
					case(1,2,3);	pm2 = 1.0d0
					case(4);	pm2 = 1.0d0 / sqrt(4.0d0*pi)
				end select

				coef0 = coef0 + cilm(1,1,1) * pm2
				coef0s = coef0s + cilm(1,1,1) * pm2 	! symsign is always 1 for l=m=0
				
				k = 2
				pm1 =  f1(k) * z * pm2
				coef0 = coef0 + cilm(1,2,1) * pm1
				coef0s = coef0s + cilm(1,2,1) * pm1 * symsign(k)
				
				do l=2, lmax_comp, 1
					l1 = l+1
					k = k+l
					p = f1(k) * z * pm1 - f2(k) * pm2
					coef0 = coef0 + cilm(1,l1,1) * p
					coef0s = coef0s + cilm(1,l1,1) * p * symsign(k)
					pm2 = pm1
					pm1 = p
				enddo
				
				select case(lnorm)
					case(1,2);	pmm = sqr(2) * scalef
					case(3);	pmm = scalef
					case(4);	pmm = sqr(2) * scalef / sqrt(4.0d0*pi)
				end select
				
				rescalem = 1.0d0/scalef
				kstart = 1
			
				do m = 1, lmax_comp-1, 1
				
					m1 = m+1
					rescalem = rescalem * u
					kstart = kstart+m+1
					
					select case(lnorm)
						case(1,4)
							pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
							pm2 = pmm
						case(2)
							pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
							pm2 = pmm / sqr(2*m+1)
						case(3)
							pmm = phase * pmm * dble(2*m-1)
							pm2 = pmm
					end select
					
					coef(m1) = coef(m1) + dcmplx(cilm(1,m1,m1), - cilm(2,m1,m1)) * pm2
					coefs(m1) = coefs(m1) + dcmplx(cilm(1,m1,m1), - cilm(2,m1,m1)) * pm2
					! symsign(kstart) = 1
										
					k = kstart+m+1
	   				pm1 = z * f1(k) * pm2
	   				
	   				coef(m1) = coef(m1) + dcmplx(cilm(1,m1+1,m1), - cilm(2,m1+1,m1)) * pm1		
					coefs(m1) = coefs(m1) - dcmplx(cilm(1,m1+1,m1), - cilm(2,m1+1,m1)) * pm1
					! symsign = -1
	   				
					do l = m+2, lmax_comp, 1
						l1 = l+1
						k = k + l
						p = z * f1(k) * pm1 - f2(k) * pm2
						pm2 = pm1
                  				pm1 = p
						coef(m1) = coef(m1) + dcmplx(cilm(1,l1,m1), - cilm(2,l1,m1)) * p
						coefs(m1) = coefs(m1) + dcmplx(cilm(1,l1,m1), - cilm(2,l1,m1)) * p * symsign(k)
					enddo
					
					coef(m1) = coef(m1) * rescalem
					coefs(m1) = coefs(m1) * rescalem
					
				enddo			
								
				rescalem = rescalem * u
				
          			select case(lnorm)
            				case(1,4);	pmm = phase * pmm * sqr(2*lmax_comp+1) / sqr(2*lmax_comp) * rescalem
            				case(2);	pmm = phase * pmm / sqr(2*lmax_comp) * rescalem
            				case(3);	pmm = phase * pmm * dble(2*lmax_comp-1) * rescalem
        			end select
         			
        			coef(lmax_comp+1) = coef(lmax_comp+1) + dcmplx(cilm(1,lmax_comp+1,lmax_comp+1), &
							- cilm(2,lmax_comp+1,lmax_comp+1)) * pmm
				coefs(lmax_comp+1) = coefs(lmax_comp+1) + dcmplx(cilm(1,lmax_comp+1,lmax_comp+1), &
							- cilm(2,lmax_comp+1,lmax_comp+1)) * pmm
				! symsign = 1
		
				coef(1) = dcmplx(coef0,0.0d0)
				coef(2:lmax_comp+1) = coef(2:lmax_comp+1)/2.0d0
               			
               			call dfftw_execute(plan)	! take fourier transform
               			gridglq(i,1:nlong) = grid(1:nlong)
               			
               			coef(1) = dcmplx(coef0s,0.0d0)
				coef(2:lmax_comp+1) = coefs(2:lmax_comp+1)/2.0d0
               			
               			call dfftw_execute(plan)	! take fourier transform
               			gridglq(i_s,1:nlong) = grid(1:nlong)
 

			endif
			
		enddo
		
	endif
	
	call dfftw_destroy_plan(plan)
				
end subroutine MakeGridGLQ

